### DISCUSSION TITLE
Plan Quality

### DISCUSSION THREAD

**Discussion by mikand** [aiplan4eu/unified-planning#69] [Ideas]
The topic of this discussion is how to model in a general, expressive yet usable and solvable way the metrics for plan quality.

Plan quality is important in several use-cases and also for RL integration (scikit-decide).

First, here is a list of plan quality kinds we want to express:
- Action costs: both state-independent and state-dependent
- Numerical metric to be evaluated on final state (e.g. final battery, linear combination of fluents in goal state)
- Makespan
- Number of achieved goals (a-la oversubscription planning) 
- Soft-goals (possibly weighted)
- Soft state trajectory constraints
- State-specific cost (i.e. specify a function `cost: States -> Reals`) 
  The semantics of this is to be defined. For example: given a plan `s1...sn` the overall cost is `sum_{si \in {s1...sn}} cost(s_i)`
  Note: possible relation with PDDL axioms

Second, we might want to combine multiple goals to define a multi-objective optimization problem.
- Lexicographic
- Pareto

Here is a minimal proposal to represent optimization. To be discussed.
```python
class PlanQualityMetric():
    '''This is the base class of any metric for plan quality'''
    pass

class MinimizeMakespan(PlanQualityMetric):
    pass

class MinimizeExpressionOnFinalState(PlanQualityMetric):
    def __init__(self, expression:upf.model.FNode):
        self.expression = expression

# Other basic metrics as needed

# To use a plan metric we add it to the problem definition
problem = upf.Problem('test')
x = upf.model.Fluent('x', upf.RealType(0, 100))
[...]
problem.add_quality_metric(MinimizeExpressionOnFinalState(x))

#If we add more than one optimality metric we need to specify the multi-objective policy
problem.add_quality_metric(MinimizeMakespan())
problem.set_muilti_objective_policy('pareto') # Either pareto or lexicographic 
```

**Comment by mikand** [comment]
Here are some more code snippets to be discussed.

**Action costs**
```python
problem.set_action_cost(a, cost)
...
problem.add_quality_metric(MinimizeActionCosts())
```
or 
```python
act.set_cost(cost)
...
problem.add_quality_metric(MinimizeActionCosts())
```
in both cases, cost is an expression and can be either:
- a constant (fixed action costs)
- an expression depending only on action parameters (when grounding, this becomes a constant)
- an expression containing action fluents (state-dependent cost)

Note that this is equivalent, but simpler and cleaner to adding a fresh fluent "total-cost" and adding a numeric effect in all actions incrementing it and requiring to `MinimizeExpressionOnFinalState(total_cost)` that to my understanding is the PDDL way of expressing action costs.

**Satisficing vs Optimal Planning**
The idea is to let the user express in the planner selection queries whether she wants to deal with optimality at all and whether a real optimal solution is required or not.
```python
with OneshotPlanner(kind=problem.kind(), optimality_guarantee=None) as s: # ignore optimality entirely, any planner can be selected
with OneshotPlanner(kind=problem.kind(), optimality_guarantee=SATISFICING) as s: # satisficing planner, only optimal or satisficing planners are eligible
with OneshotPlanner(kind=problem.kind(), optimality_guarantee=OPTIMAL) as s: # guaranteed optimal planner
```
If a planner is selected by name, a warning is issued if it is not guaranteed to be optimal:
```python
with OneshotPlanner(name="non-optimal-planner") as s: #reports a warning "Warning: non-optimal-planner is not guaranteed to return an optimal solution for the given problem encompassing plan quality metrics"
```
the warning can be silenced by explicitly stating that the user does not care about plan quality:
```python
with OneshotPlanner(name="non-optimal-planner", optimality_guarantee=None) as s:
```

**Measuring plan quality**
Given a problem with plan quality metrics, validators that declare to be capable of dealing with the problem shall be able to return the quality metrics for a given plan
```python
makespan = MinimizeMakespan()
problem.add_quality_metric(makespan)
with PlanValidator(kind=problem.kind()) as v:
    result = v.validate(plan)
    assert result.is_valid()
    print(result.get_quality_metric_value(makespan))    
```

**Multi-objective weighted sum**
Maybe it could be useful to attach a weight to each quality metric and offer a to optimize the linear combination as an alternative to proper multi-objective optimization. However, I am not aware of any planner that can solve this in general.

```python
problem = upf.Problem('test')
x = upf.model.Fluent('x', upf.RealType(0, 100))
[...]
problem.add_quality_metric(MinimizeExpressionOnFinalState(x), weight=3)

#If we add more than one optimality metric we need to specify the multi-objective policy
problem.add_quality_metric(MinimizeMakespan(), weight=9)
problem.set_muilti_objective_policy('linear') # Minimizes 3*final_value(x) + 9*makespan() 
``` 

**Notes**
Note that with a change like the one discussed we will need to enrich/edit our taxonomy (https://github.com/aiplan4eu/upf/wiki/Planning-Taxonomy) and the ProblemKind class so that the UPF can distinguish and predicate on whether a problem requires optimization and of which kind.

For example:
```python
k = problem.get_kind()
# Possible queries
k.has_optimization()
k.has_multi_objective_optimization()
k.has_makespan_minimization()
k.has_constant_action_costs()
```

**Comment by arbimo** [comment]
When supporting optimizing solvers it is very important to give the user some feedback regarding the current best solution: 90% of the time the user will abort the search before an optimal solution is found.

One powerful way to do this is to let the user have access to intermediate solutions with a callback.
Here we could let the user provide an optional callback function `on_new_plan` as an optional parameter to solve:

```py
# Solve the problem, printing out every intermediate plan found
solver.solve(on_new_plan=lambda plan: print(plan))
```

Interestingly, such a mechanism also allows the support of anytime planning.
If we check the return value of the callback, we can also use it as a control mechanism to stop the search:

```py
# Callback to call when a new plan is found (needed because python does not support multiline lambdas)
def callback(plan):
  print(f"Found plan with cost {plan.cost}")
  if plan.cost < 5:
    print("I am happy now, we can stop")
    return false  # Stop searching
  else:
    return true # continue searching

planner.solve(on_new_plan=callback)
```

This is easy to support on the API side. In terms of integration on the other hand, it requires that the planner gives an access to the intermediate solution it finds. I don't know of any standard way to do this : some planners just print the current plan/cost, others write every new plan to a file based on a naming convention (`plan.1.txt`, `plan.2.txt`, ...).

The good thing is that it is not a problem if the planner does not hand out the intermediate solutions. It simply means that callback will never be called, negatively impacting the user experience but not breaking any contract.

**Reply by mikand** [reply]
I really like the idea!

**Comment by madmage** [comment]
What is the difference between a SATISFICING plan and a plan that ignores the optimality (None)?
A solution that satisfies some conditions, isn't a solution that satisfies a goal?
